core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -426,18 +427,41 @@ trait DataSkippingReaderBase
       }
     case EqualTo(v: Literal, a) =>
       constructDataFilters(EqualTo(a, v))
+    case EqualNullSafe(v: Literal, a) =>
+      constructDataFilters(EqualNullSafe(a, v))
 
     // Match any file whose min/max range contains anything other than the rejected point.
-    case Not(EqualTo(SkippingEligibleColumn(a), SkippingEligibleLiteral(v))) =>
+    case Not(EqualTo(SkippingEligibleColumn(a), lit @ SkippingEligibleLiteral(v)))
+        if lit.value != null =>
       val minCol = StatsColumn(MIN, a)
       val maxCol = StatsColumn(MAX, a)
       getStatsColumnOpt(minCol).flatMap { min =>
         getStatsColumnOpt(maxCol).map { max =>
           DataSkippingPredicate(!(min === v && max === v), minCol, maxCol)
         }
       }
+    case Not(EqualNullSafe(SkippingEligibleColumn(a), lit @ SkippingEligibleLiteral(v)))
+        if lit.value != null =>
+      val minCol = StatsColumn(MIN, a)
+      val maxCol = StatsColumn(MAX, a)
+      val nullCountCol = StatsColumn(NULL_COUNT, a)
+      // since `Not(Literal(null, _) <=> NotNullLiteral(v, _))` returns true, means we can't
+      // skipping the file which has stats with `nullCount > 0`.
+      getStatsColumnOpt(nullCountCol).flatMap { nullCount =>
+        getStatsColumnOpt(minCol).flatMap { min =>
+          getStatsColumnOpt(maxCol).map { max =>
+            DataSkippingPredicate(
+              !(min === v && max === v && nullCount === 0),
```

Comment:
although it's impossible today, could you use 
```
              min.isNull || max.isNull || nullCount.isNull ||
                !(min === v && max === v && nullCount === 0),
```
for safety?

Reply:
make sense